File: Sources/XMTP/Messages/PrivateKeyBundleV2.swift
//
//  PrivateKeyBundleV2.swift
//
//
//  Created by Pat Nakajima on 11/26/22.
//

import Foundation
import secp256k1
import XMTPProto

public typealias PrivateKeyBundleV2 = Xmtp_MessageContents_PrivateKeyBundleV2

extension PrivateKeyBundleV2 {
	func sharedSecret(peer: SignedPublicKeyBundle, myPreKey: SignedPublicKey, isRecipient: Bool) throws -> Data {
		var dh1: Data
		var dh2: Data
		var preKey: SignedPrivateKey

		if isRecipient {
			preKey = try findPreKey(myPreKey)
			dh1 = try sharedSecret(private: preKey.secp256K1.bytes, public: peer.identityKey.secp256K1Uncompressed.bytes)
			dh2 = try sharedSecret(private: identityKey.secp256K1.bytes, public: peer.preKey.secp256K1Uncompressed.bytes)
		} else {
			preKey = try findPreKey(myPreKey)
			dh1 = try sharedSecret(private: identityKey.secp256K1.bytes, public: peer.preKey.secp256K1Uncompressed.bytes)
			dh2 = try sharedSecret(private: preKey.secp256K1.bytes, public: peer.identityKey.secp256K1Uncompressed.bytes)
		}

		let dh3 = try sharedSecret(private: preKey.secp256K1.bytes, public: peer.preKey.secp256K1Uncompressed.bytes)

		let secret = dh1 + dh2 + dh3

		return secret
	}

	func sharedSecret(private privateData: Data, public publicData: Data) throws -> Data {
		let publicKey = try secp256k1.Signing.PublicKey(rawRepresentation: publicData, format: .uncompressed)

		let sharedSecret = try publicKey.multiply(privateData.bytes, format: .uncompressed)

		return sharedSecret.rawRepresentation
	}

	func findPreKey(_ myPreKey: SignedPublicKey) throws -> SignedPrivateKey {
		for preKey in preKeys {
			if preKey.matches(myPreKey) {
				return preKey
			}
		}

		throw PrivateKeyBundleError.noPreKeyFound
	}

	func toV1() throws -> PrivateKeyBundleV1 {
		var bundle = PrivateKeyBundleV1()
		bundle.identityKey = try PrivateKey(identityKey)
		bundle.preKeys = try preKeys.map { try PrivateKey($0) }
		return bundle
	}

	func getPublicKeyBundle() -> SignedPublicKeyBundle {
		var publicKeyBundle = SignedPublicKeyBundle()

		publicKeyBundle.identityKey = identityKey.publicKey
		publicKeyBundle.identityKey.signature = identityKey.publicKey.signature
		publicKeyBundle.identityKey.signature.ensureWalletSignature()
		publicKeyBundle.preKey = preKeys[0].publicKey

		return publicKeyBundle
	}
}

File: library/src/main/java/org/xmtp/android/library/messages/PrivateKeyBundleV2.kt
package org.xmtp.android.library.messages

import org.xmtp.android.library.XMTPException

typealias PrivateKeyBundleV2 = org.xmtp.proto.message.contents.PrivateKeyOuterClass.PrivateKeyBundleV2

fun PrivateKeyBundleV2.sharedSecret(
    peer: SignedPublicKeyBundle,
    myPreKey: SignedPublicKey,
    isRecipient: Boolean,
): ByteArray {
    val dh1: ByteArray
    val dh2: ByteArray
    val preKey: SignedPrivateKey
    if (isRecipient) {
        preKey = findPreKey(myPreKey)
        dh1 = this.sharedSecret(
            preKey.secp256K1.bytes.toByteArray(),
            peer.identityKey.secp256K1Uncompressed.bytes.toByteArray()
        )
        dh2 = this.sharedSecret(
            identityKey.secp256K1.bytes.toByteArray(),
            peer.preKey.secp256K1Uncompressed.bytes.toByteArray()
        )
    } else {
        preKey = findPreKey(myPreKey)
        dh1 = this.sharedSecret(
            identityKey.secp256K1.bytes.toByteArray(),
            peer.preKey.secp256K1Uncompressed.bytes.toByteArray()
        )
        dh2 = this.sharedSecret(
            preKey.secp256K1.bytes.toByteArray(),
            peer.identityKey.secp256K1Uncompressed.bytes.toByteArray()
        )
    }
    val dh3 = this.sharedSecret(
        preKey.secp256K1.bytes.toByteArray(),
        peer.preKey.secp256K1Uncompressed.bytes.toByteArray()
    )
    return dh1 + dh2 + dh3
}

@OptIn(ExperimentalUnsignedTypes::class)
fun PrivateKeyBundleV2.sharedSecret(privateData: ByteArray, publicData: ByteArray): ByteArray {
    return uniffi.xmtpv3.diffieHellmanK256(privateData, publicData).toUByteArray().toByteArray()
}

fun PrivateKeyBundleV2.findPreKey(myPreKey: SignedPublicKey): SignedPrivateKey {
    for (preKey in preKeysList) {
        if (preKey.matches(myPreKey)) {
            return preKey
        }
    }
    throw XMTPException("No Pre key set")
}

fun PrivateKeyBundleV2.toV1(): PrivateKeyBundleV1 {
    return PrivateKeyBundleV1.newBuilder().also {
        it.identityKey = PrivateKeyBuilder.buildFromSignedPrivateKey(identityKey)
        it.addAllPreKeys(preKeysList.map { key -> PrivateKeyBuilder.buildFromSignedPrivateKey(key) })
    }.build()
}

fun PrivateKeyBundleV2.getPublicKeyBundle(): SignedPublicKeyBundle {
    return SignedPublicKeyBundle.newBuilder().also {
        it.identityKey = identityKey.publicKey
        it.identityKey = it.identityKey.toBuilder().also { idKeyBuilder ->
            idKeyBuilder.signature = identityKey.publicKey.signature.ensureWalletSignature()
        }.build()
        it.preKey = preKeysList[0].publicKey
    }.build()
}
